retrofit/src/main/java/retrofit/converter/GsonConverter.java

```diff
@@ -45,69 +46,38 @@ public GsonConverter() {
    * decoding from JSON (when no charset is specified by a header) will use UTF-8.
    */
   public GsonConverter(Gson gson) {
-    this(gson, "UTF-8");
+    this(gson, Charset.forName("UTF-8"));
   }
 
   /**
    * Create an instance using the supplied {@link Gson} object for conversion. Encoding to JSON and
    * decoding from JSON (when no charset is specified by a header) will use the specified charset.
    */
-  public GsonConverter(Gson gson, String charset) {
+  public GsonConverter(Gson gson, Charset charset) {
     this.gson = gson;
     this.charset = charset;
+    this.mediaType = MediaType.parse("application/json; charset=" + charset.name());
   }
 
-  @Override public Object fromBody(TypedInput body, Type type) throws IOException {
-    String charset = this.charset;
-    if (body.mimeType() != null) {
-      charset = MimeUtil.parseCharset(body.mimeType(), charset);
+  @Override public Object fromBody(ResponseBody body, Type type) throws IOException {
+    Charset charset = this.charset;
+    if (body.contentType() != null) {
+      charset = body.contentType().charset(charset);
     }
-    InputStreamReader isr = null;
+
+    InputStream is = body.byteStream();
     try {
-      isr = new InputStreamReader(body.in(), charset);
-      return gson.fromJson(isr, type);
+      return gson.fromJson(new InputStreamReader(is, charset), type);
```

Comment:
Should OkHttp have an overload for `body.charStream` that takes a default charset? We should.